modules/*: trust CA certificates on the nodes

s-urbaniak opened the pull request:
This also bumps the ignition provider to 1.0.0 to support S3 downloads
via ignition.

Fixes #INST-67

Commit:
modules/*: trust CA certificates on the nodes

This also bumps the ignition provider to 1.0.0 to support S3 downloads
via ignition.

Fixes #INST-67

Commit:
Documentation/examples: regenerate

s-urbaniak commented:
@squat PTAL

squat commented on modules/gcp/worker-igm/ignition.tf:
```diff
@@ -3,6 +3,7 @@ data "ignition_config" "main" {
+    "${var.ign_ca_cert_id_list}",
   ]
 
   systemd = [
```
The worker is missing the `var.ign_update_ca_certificates_dropin_id` service

s-urbaniak replied:
fixed, thanks!

squat commented on platforms/metal/cl/bootkube-controller.yaml.tmpl:
```diff
@@ -38,6 +38,11 @@ systemd:
     - name: bootkube.service
       enable: false
       contents: {{.ign_bootkube_service_json}}
+    - name: update-ca-certificates.service
```
not for this pr but we really need to refactor this platform so that it uses the same ignition format as all the other platforms

s-urbaniak replied:
totally agreed 👍

squat reviewed (changes requested):
Looks great overall. I found one issue that I think breaks GCP. PTAL

squat commented:
ok to test

s-urbaniak commented:
/cc @sym3tri @squat 

this will need to slip unfortunately :-( despite the countermeasures in #2422 this exceeds the userdata limit on AWS. We need to move non-ciritical (security wise) assets to S3 to free up space in userdata unfortunately. This implies a larger refactoring though.

Commit:
modules/ignition: unify etcd certificate generation

Currently etcd certificate ignition files are created in their
respective platform module.

This unifies it by declaring the ignition file units centrally in the
ignition module.

Commit:
modules/tls/etcd: remove zip generation

Currently the etcd TLS module also generates a zip file which is only
used on AWS to reduce the userdata size to be <20k (hard limit on AWS).

Since the etcd TLS assets will be provisioned via S3 this optimization/hack
is not needed any more.

Commit:
*/aws: use S3 for ignition provisioning

We hit the limits of the AWS userdata limit (20k) constantly.

This fixes it by provisioning a minimal ignition configuration only
which points to a replacement ignition configuration hosted on S3.

This also removes workarounds/hacks to keep the userdata size small,
especially for provisioning TLS assets on etcd nodes.

Commit:
*/azure: use unified etcd TLS ignition files

Commit:
*/gcp: use unified etcd TLS ignition files

Commit:
*/openstack: use unified etcd TLS ignition files

Commit:
*/vmware: use unified etcd TLS ignition files

s-urbaniak commented:
@squat @enxebre PTAL

s-urbaniak commented:
retest this please

squat reviewed (approved):
LGTM once green

s-urbaniak commented on modules/ignition/resources/dropins/10-always-update-ca-certificates.conf:
```diff
@@ -0,0 +1,6 @@
+
+[Service]
+ExecStart=
+ExecStart=/usr/sbin/update-ca-certificates
```
/cc @lucab This works, because we are enforcing rehashing all files. Rather the default dropin should work, but openssl is pointing to the wrong default bundle.

s-urbaniak commented:
all green, merging

sym3tri commented on modules/aws/etcd/nodes.tf:
```diff
@@ -20,14 +20,88 @@ data "aws_ami" "coreos_ami" {
+      "Action": [
+        "ecr:DescribeRepositories",
+        "ecr:ListImages",
+        "ecr:BatchGetImage"
```
@s-urbaniak Why did you need to add ECR permissions?

s-urbaniak replied:
It might've been a copy&pasta oversight, but without this the machine didn't seem to boot.

wking commented on modules/aws/etcd/ignition_s3.tf:
```diff
@@ -0,0 +1,25 @@
+resource "aws_s3_bucket_object" "ignition_etcd" {
+  count = "${length(var.external_endpoints) == 0 ? var.instance_count : 0}"
```
I'm new to Terraform and was scratching my head about this after not finding `count` in [the `aws_s3_bucket_object` docs][1].  It turns out it is a [meta-parameter][2] (more [here][3]).  I thought I'd add this note in case it helps any other Terraform newbies ;).

[1]: https://www.terraform.io/docs/providers/aws/r/s3_bucket_object.html
[2]: https://www.terraform.io/docs/configuration/resources.html#meta-parameters
[3]: https://www.terraform.io/docs/configuration/resources.html#using-variables-with-count
